python-checks/src/main/java/org/sonar/python/checks/tests/AssertOnDissimilarTypesCheck.java

```diff
@@ -0,0 +1,166 @@
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.CALL_EXPR, ctx -> {
+      CallExpression callExpression = (CallExpression) ctx.syntaxNode();
+      if (!callExpression.callee().is(Tree.Kind.QUALIFIED_EXPR)) {
+        return;
+      }
+
+      QualifiedExpression qualifiedExpression = (QualifiedExpression) callExpression.callee();
+      if (!(qualifiedExpression.qualifier().is(Tree.Kind.NAME) && ((Name) qualifiedExpression.qualifier()).name().equals("self"))) {
+        return;
+      }
+
+      checkArguments(ctx, callExpression, qualifiedExpression);
+    });
+  }
+
+  private static void checkArguments(SubscriptionContext ctx, CallExpression callExpression, QualifiedExpression qualifiedExpression) {
+    boolean isAnAssertIdentity = assertToCheckIdentity.contains(qualifiedExpression.name().name());
+    boolean isAnAssertEquality = assertToCheckEquality.contains(qualifiedExpression.name().name());
+    boolean shouldRaise = false;
+
+    if (!isAnAssertEquality && !isAnAssertIdentity) {
+      return;
+    }
+
+    ArgList args = callExpression.argumentList();
+    if (args == null) {
+      return;
+    }
+
+    List<Argument> arguments = args.arguments();
+    if (!(arguments.size() >= 2 && arguments.get(0).is(Tree.Kind.REGULAR_ARGUMENT) && arguments.get(1).is(Tree.Kind.REGULAR_ARGUMENT))) {
+      return;
+    }
+
+    Expression left = ((RegularArgument) arguments.get(0)).expression();
+    Expression right = ((RegularArgument) arguments.get(1)).expression();
+    if (!checkArgumentsCantBeIdentical(left, right)) {
+      return;
+    } else if (isAnAssertIdentity) {
+      shouldRaise = true;
+    }
+    if (isAnAssertEquality && checkArgumentsCantBeEqual(left, right)) {
+      shouldRaise = true;
+    }
+
+    if (shouldRaise) {
+      PreciseIssue issue = ctx.addIssue(args, message(left, right));
+      getLastAssignment(left).ifPresent(assign -> issue.secondary(assign, String.format(MESSAGE_SECONDARY, ((Name)left).name())));
+      getLastAssignment(right).ifPresent(assign -> issue.secondary(assign, String.format(MESSAGE_SECONDARY, ((Name)right).name())));
+    }
+  }
+
+  private static String message(Expression left, Expression right) {
+    String leftTypeName = InferredTypes.typeName(left.type());
+    String rightTypeName = InferredTypes.typeName(right.type());
+    String message = MESSAGE;
+    if (leftTypeName != null && rightTypeName != null) {
+      message += " (" + leftTypeName + " and " + rightTypeName + ")";
+    }
+    message += ".";
+    return message;
+  }
+
+  private static Optional<AssignmentStatement> getLastAssignment(Expression expr) {
+    if (!expr.is(Tree.Kind.NAME)) {
+      return Optional.empty();
+    }
+
+    Symbol symbol = ((Name) expr).symbol();
+    if (symbol == null) {
+      return Optional.empty();
+    }
+
+    Usage lastAssignment = symbol.usages().stream()
+      .takeWhile(usage -> usage != ((Name) expr).usage())
```

Comment:
I'm not sure I understand the `takeWhile`. Why not have the condition `usage != ((Name) expr).usage()` as part of the filter as well?

Reply:
As discussed takeWhile is different from filter in the sense that it stop at the first evaluation which return false instead of filtering it out and pursuing.
I added a test case where a variable is assigned after the self.assert statement, in this case we ensure that the last assignment before the self.assert statement is put in secondary and not the one after.

Reply:
So again, as discussed adding the use case clearly showed that this approach was not working : usages() does not list usage by order.
I added a sorted() based on the line number of the token associated to the usage, seems to do the trick.